### PR TITLE
MM-23891: Add proxy support for android and iOS

### PR DESCRIPTION
**Pull request by agnivade**
#### Summary

The android client uses the default transport, so it's easy
enough to override that with the HTTP(S)_PROXY env var.

But the iOS client has it's own transport, so we need to jump
some hoops to override that.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-23891

### CODE REVIEW

File: config/mattermost-push-proxy.json [hunk 1]
```diff
@@ -4,6 +4,7 @@
     "ThrottleMemoryStoreSize":50000,
     "ThrottleVaryByHeader":"X-Forwarded-For",
     "EnableMetrics": false,
+    "ProxyServer": "",
```
Comment (grundleborg) [inline config/mattermost-push-proxy.json:7]: Just curious, do we actually want an explicit config setting here, or is just obeying `HTTP_PROXY`/`HTTPS_PROXY` environment variables sufficient? I'm not actually sure what we do in the server, but I guess it makes the most sense to follow the same pattern here.

File: server/apple_notification_server.go [hunk 2]
```diff
@@ -38,6 +42,32 @@ func (me *AppleNotificationServer) Initialize() bool {
...
+			}
+
+			transport := &http.Transport{
+				TLSClientConfig: tlsConfig,
+				Proxy: func(request *http.Request) (*url.URL, error) {
+					return url.Parse(CfgPP.ProxyServer)
+				},
+				IdleConnTimeout: apns.HTTPClientTimeout,
+			}
+			err := http2.ConfigureTransport(transport)
```
Comment (grundleborg) [inline server/apple_notification_server.go:62]: Am I understanding correctly that Apple requires HTTP 2 for sending the push notifications to their servers?

Out of band from this PR, but we may need to check if $BIG_ENTERPRISE_CUSTOMER's HTTP proxy is compatible with HTTP 2.

File: server/server.go [hunk 3]
```diff
@@ -37,6 +38,15 @@ var gracefulServer *graceful.Server
 func Start() {
 	LogInfo("Push proxy server is initializing...")
 
+	if CfgPP.ProxyServer != "" {
```
Comment (grundleborg) [inline server/server.go:41]: Does this mean that Android already obeys the `HTTP_PROXY`/`HTTPS_PROXY` environment variables?

File: server/apple_notification_server.go (hunk 2 above)
Comment (agnivade) [reply server/apple_notification_server.go:62]: Correct.

> Out of band from this PR, but we may need to check if $BIG_ENTERPRISE_CUSTOMER's HTTP proxy is compatible with HTTP 2.

Yes, I have talked with Stu, and the $BIG_CUSTOMER will move to a new proxy which supports HTTP/2. They are aware of this.

**Comment by agnivade** [comment]
@grundleborg - Let me explain the answers to 1st and 3rd questions here. I already attempted to do that in the PR message, but perhaps it wasn't clear. :smiley: 

Yes, Android already obeys the environment variables. And yes, we need an explicit config variable.

The reason is because all Go http clients have a default transport layer. Unless somebody overrides that, one can just set the environment vars and the net/http package will transparently apply the proxy to all requests. Which is what we do in the MM server.

But .. the apns client uses it's own transport layer. And hence, we need to explicitly override it. And hence, we need to set both the environment var and explicitly override the transport. We can always get the value from the env var, and override the transport. But I felt that setting an env var, and setting a config property is kinda one and the same. And since setting the env var, doesn't affect apns anyways, so thought of using the config variable.

**Comment by grundleborg** [comment]
Thanks for the explanation, I definitely get it now :)

I think we should be consistent with mattermost-server, where it appears that we honour the environment variable rather than providing a config setting. I'm 1/5 so open to being persuaded otherwise though.

**Comment by agnivade** [comment]
Ah, no worries. Makes sense to me.

**Comment by agnivade** [comment]
@srkgupta - This is ready for testing now. Let me know whenever you are ready, I can give you the lowdown on things.

**Comment by agnivade** [comment]
Awesome work !

**Pull request merged**